File: Gombang/[LeetCode75] - [DFS]/872_Leaf-Similar_Trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<int> root1Vec;
        vector<int> root2Vec;

        DFS(root1, root1Vec);
        DFS(root2, root2Vec);

        if (root1Vec.size() != root2Vec.size())
            return false;

        for (int i = 0; i < root1Vec.size(); i++)
        {
            if (root1Vec[i] != root2Vec[i])
                return false;
        }

        return true;
    }

    void DFS(TreeNode* node, vector<int>& vec)
    {
        if (node == nullptr)
            return;

        if (node->left == nullptr && node->right == nullptr)
        {
            vec.push_back(node->val);
            return;
        }

        DFS(node->left, vec);
        DFS(node->right, vec);
    }
};

File: Seori/LeetCode 75/Binary Tree_DFS/872. Leaf-Similar Trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        def dfs(root: Optional[TreeNode], leaf_value_sequence: List) -> None:
            # 재귀 종료 조건 : 리프에 도착했을 때
            if root == None: 
                return
            
            # 문제 조건 : 자식이 없는 리프노드의 경우
            if root.left == None and root.right == None:
                leaf_value_sequence.append(root.val)

            # 트리 전위순회
            left = dfs(root.left, leaf_value_sequence)
            right = dfs(root.right, leaf_value_sequence)
            return leaf_value_sequence
            
        return dfs(root1, []) == dfs(root2, [])
